### PR TITLE
Python Rewrite

### PR DESCRIPTION
**Pull request by mikepurvis**
The basic idea of the package is unchanged, and the new implementation of the `install` script is intended to be a drop-in replacement for the bash-based one, so the behaviours described on the [ROS Wiki](http://wiki.ros.org/robot_upstart) still very much apply. However, the Python implementation has enabled creating a much improved Code API, and properly documenting it, which has been done using Sphinx. ~~I've placed a copy of the generated docs online here: http://mikepurvis.github.io/robot_upstart/~~ Now that this is merged, see the docs here: http://docs.ros.org/api/robot_upstart/html/

Summary of changes:
- The templated system files are now created using empy, the tool used in catkin and by the ROS message generators.
- The new Job class captures the user's desired configuration, without making any changes.
- The specifics of creating an _upstart_ job have been moved to `providers.Upstart`, potentially facilitating other backends for other systems.
- All logic now executes in the unprivileged process; only a very simple `create_files` shim is executed as root (it's passed a pickled recipe of what exactly to do), and there's a clear statement of which files it intends to create before it does anything (no more failed `pushd` calls spewing files everywhere).
- No tests as yet, but this architecture is much more testable.
- The install script is now Python, and a pass-thru to the Job class.
- The upstart config is now set to log stdout/stderr to the `/var/log/upstart/JOB.log` file. Maybe remove the syslog logging now?
- The ROS_MASTER_URI now defaults to 127.0.0.1:11311, rather than using the ROS_IP value.

The new implementation has been tested with the [present version of Husky's install script](https://github.com/husky/husky_bringup/blob/indigo-devel/scripts/install), but I'd like to get some additional checking before releasing these changes into Indigo, since any corner cases could be an unpleasant surprise to someone not aware of these changes.

Once this is actually merged and released, Husky and the other CPR products will begin transitioning their install scripts to use the Python API.

Provisos:
- the `uninstall` command is not implemented. This is in part because I'd like some feedback on what it should do— should it be up to the Provider to simply assert which files would have been created and rm those? Or should it create a dotfile or something with a log of files _actually_ created, and remove those? Or something else?
- the start/stop scripts could be replaced by a simpler Python wrapper which figures out and sets up the environment, but that can be a separate change from this one.

Addresses #5, #7. Please review: @paulbovbel, @akalmbach, @ipa-fmw, @ipa-mig, @BennyRe, @girtslinde

### CODE REVIEW

**Comment by BennyRe** [comment]
Works for me like the old one. _thumbs up_

**Comment by girtslinde** [comment]
Tried on the robot, everything still works. Looks good! :)

**Comment by paulbovbel** [comment]
As discussed, please add a --wait option for jobs which do not bring up master, and a purge parameter/interface for removing launch files during uninstall.

Looks good :+1:

File: templates/job.conf.em [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+description "bringup @(name)"
+
+@[if interface]@
+@# Start and stop ROS with the availability of the network interface
+@# which is used as the source for ROS_IP.
+start on net-device-up IFACE=@(interface)
+stop on net-device-down IFACE=@(interface)
+@[else]@
+@# Start and stop ROS with the system.
+start on startup
```
Comment (mikepurvis) [inline templates/job.conf.em]: This is no good— catkin's setup.bash invocation requires momentary use of `/tmp`, which isn't available right at boot. I believe the correct trigger here would be `local-filesystems`.

cf. http://upstart.ubuntu.com/cookbook/#mountall-debian-and-ubuntu-specific

**Comment by mikepurvis** [comment]
Okay, this is merging now. Further concerns will be addressed as bugs.

**Pull request merged**